src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp

```diff
@@ -0,0 +1,165 @@
+    const double total_energy_density, const double momentum_density_squared,
+    const double momentum_density_dot_magnetic_field,
+    const double magnetic_field_squared,
+    const double rest_mass_density_times_lorentz_factor,
+    const EquationsOfState::EquationOfState<true, ThermodynamicDim>&
+        equation_of_state) noexcept {
+  // constant in cubic equation  f(eps) = eps^3 - a eps^2 + d
+  // whose root is being found at each point in the iteration below
+  const double d_in_cubic =
+      0.5 * (momentum_density_squared * magnetic_field_squared -
+             square(momentum_density_dot_magnetic_field));
+  if (0.0 > d_in_cubic) {
+    ERROR("Newman Hamlin inversion: unphysical d.");
+  }
+
+  const double initial_guess_for_pressure{0.0};
+  // bound needed so cubic equation has a positive root
+  const double minimum_pressure = cbrt(6.75 * d_in_cubic) -
+                                  total_energy_density -
+                                  0.5 * magnetic_field_squared;
+  double current_pressure =
+      std::max(minimum_pressure, initial_guess_for_pressure);
+  double previous_pressure{std::numeric_limits<double>::signaling_NaN()};
+  size_t iteration_step{0};
+  std::vector<double> aitken_pressure;
+  aitken_pressure.emplace_back(current_pressure);
+  bool converged = false;
+
+  while (true) {  // will break when relative pressure change is < 1e-10
+    if (max_iterations_ == iteration_step and not converged) {
+      return boost::none;
+    }
+
+    ++iteration_step;
+    previous_pressure = current_pressure;
+    // enforces NH Eq.(5.9): d <= (4/27) a^3 so cubic has positive root
+    current_pressure = std::max(current_pressure, minimum_pressure);
+    const double a_in_cubic =
+        total_energy_density + current_pressure + 0.5 * magnetic_field_squared;
+
+    ASSERT(a_in_cubic >= 0.0,
+           "Newman Hamlin inversion:  unphysical a = " << a_in_cubic);
+
+    // NH Eq. (5.10): d = (4/27) a^3 cos^2(phi)
+    const double phi = acos(sqrt(6.75 * d_in_cubic / cube(a_in_cubic)));
+    // NH Eq. (5.11) with l=1 is desired positive root
+    const double root_of_cubic =
+        (a_in_cubic / 3.0) * (1.0 - 2.0 * cos((2.0 / 3.0) * (M_PI + phi)));
+    // NH Eq. (5.5) with their script L being rho_h_w_squared
+    // where rho is rest_mass_density, h is specific_enthalpy,
+    // and w is the lorentz factor
+    const double rho_h_w_squared = root_of_cubic - magnetic_field_squared;
+
+    ASSERT(rho_h_w_squared > 0.0,
+           "Newman Hamlin inversion:  unphysical rho_h_w_squared = "
+               << rho_h_w_squared);
+    // NH Eq. (5.2) with (5.5) substituted in denominator
+    const double v_squared =
+        (momentum_density_squared * square(rho_h_w_squared) +
+         square(momentum_density_dot_magnetic_field) *
+             (magnetic_field_squared + 2.0 * rho_h_w_squared)) /
+        square(rho_h_w_squared * root_of_cubic);
+
+    ASSERT(0.0 <= v_squared and v_squared <= 1.0,
```

Comment:
I think this better be v_squared strictly < 1.0, otherwise next statement will fail.